purchases/src/main/kotlin/com/revenuecat/purchases/common/events/EventsManager.kt

```diff
@@ -0,0 +1,187 @@
+            )
+        }
+
+        fun paywalls(fileHelper: FileHelper): EventsFileHelper<PaywallStoredEvent> {
+            return EventsFileHelper(
+                fileHelper,
+                EventsManager.PAYWALL_EVENTS_FILE_PATH,
+                PaywallStoredEvent::toString,
+                PaywallStoredEvent::fromString,
+            )
+        }
+    }
+
+    @get:Synchronized
+    @set:Synchronized
+    private var flushInProgress = false
+
+    @get:Synchronized
+    @set:Synchronized
+    private var legacyFlushDone = false
+
+    @OptIn(ExperimentalPreviewRevenueCatPurchasesAPI::class)
+    @Synchronized
+    fun track(event: FeatureEvent) {
+        enqueue {
+            debugLog("Tracking event: $event")
+
+            val backendEvent = when {
+                event.paywallEvent != null -> event.paywallEvent!!.toBackendEvent(identityManager.currentAppUserID)
+                else -> null
+            }
+
+            if (backendEvent != null) {
+                fileHelper.appendEvent(backendEvent)
+            } else {
+                debugLog("Backend event not implemented for: $event")
+            }
+        }
+    }
+
+    @Synchronized
+    fun flushEvents() {
+        enqueue {
+            if (flushInProgress) {
+                debugLog("Flush already in progress.")
+                return@enqueue
+            }
+            flushInProgress = true
+
+            if (!legacyFlushDone) {
+                flushLegacyEvents()
+                legacyFlushDone = true
+            }
+
+            val storedEventsWithNullValues = getStoredEvents()
+            val storedEvents = storedEventsWithNullValues.filterNotNull()
+
+            if (storedEvents.isEmpty()) {
+                verboseLog("No new events to sync.")
+                flushInProgress = false
+                return@enqueue
+            }
+
+            verboseLog("New event flush: posting ${storedEvents.size} events.")
+            postEvents(
+                EventRequest(storedEvents),
+                {
+                    verboseLog("New event flush: success.")
+                    enqueue {
+                        fileHelper.clear(storedEventsWithNullValues.size)
+                        flushInProgress = false
+                    }
+                },
+                { error, shouldMarkAsSynced ->
+                    errorLog("New event flush error: $error.")
+                    enqueue {
+                        if (shouldMarkAsSynced) {
+                            fileHelper.clear(storedEventsWithNullValues.size)
+                        }
+                        flushInProgress = false
+                    }
+                },
+            )
+        }
+    }
+
+    private fun flushLegacyEvents() {
+        enqueue {
+            val storedLegacyEventsWithNullValues = getLegacyStoredEvents()
+            val storedLegacyEvents = storedLegacyEventsWithNullValues.filterNotNull()
```

Comment:
`getLegacyStoredEvents` is already returning not null values, so this is not necessary